### DISCUSSION TITLE
Guide: Unit Tests on Forked Mainnet (Lesson 10)

### DISCUSSION THREAD

**Discussion by ZeroEkkusu** [smartcontractkit/full-blockchain-solidity-course-py#668] [Show and tell]
Gm, fellow devs,

In Lesson 10, we have unit tests for our Python scripts.

If you tried making your own, though, some would fail because the state of the chain would change with each test.

Here's an example:
```
Test 1:
> Mint 1 WETH
> Assert you have 1 WETH
✔️ PASSED
```
```
Test 2:
> Mint 1 WETH
> Spend 1 WETH
> Assert you have 0 WETH
❌ FAILED - you have 1 WETH, not 0 WETH
```
That's because `Test 1` changed your WETH balance to 1 and you started `Test 2` with that 1 WETH. In other words, the chain did not reset before running `Test 2`.

Here's how to do unit tests in the `mainnet-fork-dev` environment:
- Make the `conftest.py` file in the `tests` directory
- Add the following
```python
import pytest
from scripts.helpful_scripts import LOCAL_BLOCKCHAIN_ENVIRONMENTS
from brownie import network, chain


@pytest.fixture(scope="module")
def test_only_locally():
    if network.show_active() not in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        pytest.skip()
    else:
        chain.reset()


@pytest.fixture(scope="module")
def dont_test_locally():
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        pytest.skip()
```
- Now, pass the fixtures to your tests - some examples:
```python
def test_mint_weth(test_only_locally):
    # ...
```
```python
def test_basic_integration(dont_test_locally):
    # ...
```
- Test!
```bash
brownie test --network mainnet-fork-dev
```
```bash
brownie test --network kovan
```
### Explanation
- A `@pytest.fixture` is something that will run prior to (and/or after) the test you pass it to
- `scope="module"` means that, if passed to a test, it will run once per file (module) - I keep my unit tests in separate files inside the `tests/unit` directory. Use `scope="function"` if you keep them all in a single file
- `chain.reset()` takes a snapshot before running the test and reverts to it when the test is done - ensuring the initial conditions are always the same
- As you can see, we have our network checks, as well, so, no need to do them in our tests

Btw, you can do the "Arrange" part of your test like this, too.
More info [here](https://eth-brownie.readthedocs.io/en/stable/tests-pytest-intro.html?).

HNY! 🥳

**Comment by cromewar** [comment]
This is an amazing contribution for the repository, thanks for this!!

**Comment by PatrickAlphaC** [comment]
Awesome work here!! Love it :D
